PLFM-9300 - Grid validation handles optional values

nickgros opened the pull request:
- Update constant value DB and DTO representations to discern between `null` and `undefined`
- Extend grid-index-view-template.sql to generate a JSON object document for each row in the grid, which properly includes `null` values and omits `undefined` values. JSON document is used for validation
- Update PatchRowHandler to consider required columns when creating empty values. If a column is `required`, then cells without data will be instantiated with `null` values. Otherwise, the cells will be instantiated with an `undefined` value, which is the appropriate value for an optional property.

nickgros committed:
Grid validation handles optional values

- Update constant value DB and DTO representations to discern between `null` and `undefined`
- Extend grid-index-view-template.sql to generate a JSON object document for each row in the grid, which properly includes `null` values and omits `undefined` values. JSON document is used for validation
- Update PatchRowHandler to consider required columns when creating empty values. If a column is `required`, then cells without data will be instantiated with `null` values. Otherwise, the cells will be instantiated with an `undefined` value, which is the appropriate value for an optional property.

nickgros committed:
fix quadratic complexity in index lookup

suggested by copilot

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/Synapse-Repository-Services into PLFM-9300

nickgros committed:
Handle `null` vs undefined in agent GridUpdateRequestHandler

Additionally, use ConValue everywhere to replace JSONArray representations of CRDT constant values. JSONArray cannot represent the UNDEFINED constant value.

nickgros commented on services/repository-managers/src/main/resources/grid/grid-index-view-template.sql:
```diff
@@ -17,45 +17,64 @@ WITH RECURSIVE D AS
+                   `name` VARCHAR(255) PATH '$'
+                   )
+       ) AS jt
+    WHERE JSON_CONTAINS_PATH(V1.VEC_VAL, 'one', CONCAT('$.c', jt.idx - 1, '.v'))), JSON_OBJECT()) AS VALS_JSON,
```
Cannot use `CONCAT` in a PATH literal

nickgros committed:
Check for key presence to clear value in GridUpdateRequestHandler

copilot suggested improvement

nickgros committed:
fix test

nickgros commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/grid/update/SetValue.json:
```diff
@@ -8,7 +8,7 @@
 		"value": {
 			"type": "object",
-			"description": "Literal JSON value to assign to the column for all selected rows. Supports string, number, boolean, object, array, or null (use null to clear)"
+			"description": "Literal JSON value to assign to the column for all selected rows. Supports string, number, boolean, object, array, or null, or you may omit this property to clear the value."
```
Update existing example to set `'null'` properly, and create new example to set undefined. Verify all examples are accurate.

Check with LLM to get better wording

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/node/ConstantNode.java:
```diff
@@ -2,20 +2,29 @@
+	public ConValue getConValue() {
 		return value;
 	}
 
```
looks like setValue(ConValue va) will work.

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/agent/handler/grid/GridUpdateRequestHandler.java:
```diff
@@ -93,13 +96,26 @@ public String handleEvent(ReturnControlEvent event) throws Exception {
+				for (int i = 0; i < set.size(); i++) {
+					SetValue sv = set.get(i);
+					ConValue toAdd = new ConValue(ConType.fromValue(sv.getValue()), sv.getValue());
+					JSONObject rawSetValue = updateRequestRaw.optJSONObject("update").optJSONArray("set").optJSONObject(i);
```
only need the "set" within this inner loop.

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/agent/handler/grid/GridUpdateRequestHandler.java:
```diff
@@ -93,13 +96,26 @@ public String handleEvent(ReturnControlEvent event) throws Exception {
+					SetValue sv = set.get(i);
+					ConValue toAdd = new ConValue(ConType.fromValue(sv.getValue()), sv.getValue());
+					JSONObject rawSetValue = updateRequestRaw.optJSONObject("update").optJSONArray("set").optJSONObject(i);
+					if (rawSetValue.opt("value") == JSONObject.NULL) {
```
JSONObject.isNull()

john-hill commented on lib/lib-grid/src/test/java/org/sagebionetworks/repo/model/grid/patch/ConValueTest.java:
```diff
@@ -0,0 +1,235 @@
+        ConValue vNull = ConValue.fromCompact(new JSONArray("[null]"));
+        assertEquals(ConType.NULL, vNull.getType());
+        assertEquals(JSONObject.NULL, vNull.getValue());
+
```
can we test the toCompact for each case.

john-hill commented on lib/lib-grid-db/src/main/java/org/sagebionetworks/grid/db/GridIndexDaoImpl.java:
```diff
@@ -719,14 +719,23 @@ public <T> List<T> query(String sql, SqlParameterSource paramSource, RowMapper<T
 	@Override
 	public Optional<LogicalTimestamp> findExistingConstant(String sessionIdString, Long replicaId, String jsonValue) {
```
we can remove this code, as it is now "dead code"

john-hill commented on services/repository-managers/src/main/resources/grid/grid-index-view-template.sql:
```diff
@@ -17,45 +17,64 @@ WITH RECURSIVE D AS
 	D.NODE_REP AS AN_REP, D.NODE_SEQ AS AN_SEQ,
     O1.OBJ_REP AS RO_REP, O1.OBJ_SEQ AS RO_SEQ,
     V1.VEC_REP AS VEC_REP, V1.VEC_SEQ AS VEC_SEQ,
 	JSON_ARRAY(%s) AS VALS,
```
add a new column that is the VALS_JSON injected into this SQL (the sql is build in Java like VALS).

nickgros committed:
Simplify SQL by building an expression to pass to JSON_MERGE_PRESERVE

nickgros committed:
Remove ConstantProvider & constant caching

The json-joy JavaScript library does not support reusing constants (see https://github.com/streamich/json-joy/issues/930). Since this approach results in creating patches that are incompatible with the client library, we will remove this optimization at this time.

nickgros committed:
respond to reviewer feedback

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/Synapse-Repository-Services into PLFM-9300

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/Synapse-Repository-Services into PLFM-9300

nickgros committed:
extend CellValueFilter to support distinguishing between null & undefined

add/refine examples

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/Synapse-Repository-Services into PLFM-9300

nickgros committed:
method overload requires change to null handling

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/ConValue.java:
```diff
@@ -2,17 +2,33 @@
+import org.json.JSONTokener;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Are these the correct imports?

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/ConValue.java:
```diff
@@ -46,7 +83,86 @@ public String toString() {
+			return "null";
+		}
+		if (value instanceof String) {
+			ObjectMapper mapper = new ObjectMapper();
```
There are json.org ways of achieving this goal.  JSONWriter.valueToString() should be the same thing without the bloat.

john-hill commented on lib/lib-grid/src/test/java/org/sagebionetworks/repo/model/grid/patch/ConValueTest.java:
```diff
@@ -0,0 +1,253 @@
+
+    @Test
+    public void testStringToJsonUsesQuoting() {
+        ConValue s = new ConValue(ConType.STRING, "abc");
```
add a double quote to the string

john-hill commented on lib/lib-grid/src/test/java/org/sagebionetworks/repo/model/grid/patch/ConValueTest.java:
```diff
@@ -0,0 +1,253 @@
+        assertEquals(Long.valueOf(123), v3.getValue());
+
+        // string
+        ConValue v4 = ConValue.fromJsonString("\"hello\"");
```
include escaped quote here.

john-hill reviewed (commented):
Will need a part two.

nickgros committed:
changes from review

nickgros committed:
remove imports

nickgros committed:
use compact representation everywhere

nickgros committed:
fix tests in lib-grid

nickgros committed:
fix tests in lib-grid-db

nickgros committed:
fix tests in repository-managers grid package

john-hill commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/grid/query/CellValueFilter.json:
```diff
@@ -64,15 +64,20 @@
-			},
-			"description": "A filter used to select rows based on a specific column's cell value. Use operators like 'EQUALS' or 'LIKE' with the 'value' array for standard comparisons. For user requests about finding 'empty', 'blank', or 'null' values, use the 'IS_NULL' operator and omit the 'value' property. To find rows that are not empty, use the 'IS_NOT_NULL' operator. When using either 'LIKE' or 'NOT ...
+			"type": "object",
+			"description": "A filter used to select rows based on a specific column's cell value. Use operators like 'EQUALS' or 'LIKE' with the 'value' property for standard comparisons. For user requests about finding 'empty', 'blank', or 'null' values, use the 'IS_NULL' or `IS_UNDEFINED` operators and omit the 'value' property. To find rows that are not empty, use the 'IS_NOT_NULL' or `IS_DEFINED` oper ...
```
IS_NULL will only match null values.  IS_UNDEFINDED will only match undefined.

john-hill commented on services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/grid/internal/replica/view/GridReplicaViewManagerImplAutowireTest.java:
```diff
@@ -492,163 +531,242 @@ public void testFilterWithEachColumnTypeAndIn() throws IOException {
 		GridHeader header = gridViewManager.readHeader(sessionId, replicaId).get();
+
+		// Also write a row with a `null` value, which we cannot express via the Row model.
+		Patch patch = new Patch().setPatchId(LogicalTimestamp.newIncrement(gridIndexManger.getClock(sessionId, replicaId).get(0), 1));
```
Marco is now enforcing the rule of the container's ID must be smaller than the contained ID.

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/view/GridReplicaViewManagerImpl.java:
```diff
@@ -59,23 +60,50 @@ public class GridReplicaViewManagerImpl implements GridReplicaViewManager {
+	/**
+	 * Transforms the database representation of the list of constant nodes to the Java representation
+	 */
+	private static List<ConValue> getConstantValuesFromString(String jsonStr) {
```
will delete

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/view/GridReplicaViewManagerImpl.java:
```diff
@@ -145,16 +173,23 @@ public List<RowView> querySinglePage(GridHeader header, QueryElement query) {
+		RowMapper<RowView> mapper;
+		List<SelectColumn> selectColumns = translateSelect(header, query.getSelect());
+		List<String> columnNames = selectColumns.stream().map(SelectColumn::getColumnName).collect(Collectors.toList());
+		if (query.isAggregate()) {
```
ternary?

nickgros committed:
changes from code review

nickgros committed:
Merge branch 'develop' of https://github.com/Sage-Bionetworks/Synapse-Repository-Services into PLFM-9300

nickgros committed:
account for changes in #5436
